pkg/gui/controllers/branches_controller.go

```diff
@@ -97,9 +97,75 @@ func (self *BranchesController) GetKeybindings(opts types.KeybindingsOpts) []*ty
 			Handler:     self.checkSelectedAndReal(self.rename),
 			Description: self.c.Tr.LcRenameBranch,
 		},
+		{
+			Key:         opts.GetKey(opts.Config.Branches.SetUpstream),
+			Handler:     self.checkSelected(self.setUpstream),
+			Description: "(Un)set upstream",
+			OpensMenu:   true,
+		},
 	}
 }
 
+// TODO: this is the same as SyncController.parseUpstream
+func (self *BranchesController) parseUpstream(upstream string) (string, string, error) {
+	var upstreamBranch, upstreamRemote string
+	split := strings.Split(upstream, " ")
+	if len(split) != 2 {
+		return "", "", errors.New(self.c.Tr.InvalidUpstream)
+	}
+
+	upstreamRemote = split[0]
+	upstreamBranch = split[1]
+
+	return upstreamRemote, upstreamBranch, nil
+}
+
+func (self *BranchesController) promptForUpstream(currentBranch *models.Branch, onConfirm func(string) error) error {
+	// TODO: how to get this from `gui`?
+	// suggestedRemote := self.getSuggestedRemote()
```

Comment:
Currently SyncController receives `getSuggestedRemote` from the gui package as an argument, but given that we need to use it here I'd move it into a controller helper. I'd add a helper called UpstreamHelper which contains GetSuggestedRemote, PromptForUpstream, and ParseUpstream, and then use those methods from here like `self.helpers.Upstream.PromptForUpstream`. I'm imagining something like this for the struct:
```go
type UpstreamHelper struct {
	c        *types.HelperCommon
	model    *types.Model
	suggestionsHelper *SuggestionsHelper
}
```
We'll need suggestionsHelper there so that we can call `self.suggestionsHelper.GetRemoteBranchesSuggestionsFunc`

What do you reckon?

Reply:
I pushed the things you recommended, although with some slight changes, and with the addition that it seems a tad overtly convoluted.

I didn't pass the `suggestionsHelper` because I needed only one function (seems like a wrong choice, in hindsight), and I still needed the `gui.GetSuggestedRemote` so I passed that as funcref as well.

Help, please.

Reply:
I agree about it being a little convoluted, though at the moment nothing comes to mind as a better approach. Open to suggestions on a better organisation of controllers vs helpers vs contexts cos it's been bugging me a little

Reply:
Well it's a tad less convoluted now when `misc.go` is removed. At least considering only this helper.

As for suggestions regarding the general architecture, can't contribute to that myself, I've taken the status quo as-is, but I can ask around what other approaches are there.

Reply:
awesome :)